lib/failure_detector.ml

```diff
@@ -0,0 +1,138 @@
+open Base.Hashtbl
+
+type message =
+  | Ping
+  | Acknowledge
+  | PingRequest of Address.t
+[@@deriving bin_io]
+
+type config = {
+  protocol_period : int;
+  round_trip_time : int;
+  timeout : int;
+  mutable peers_to_ping : int;
+}
+
+type t = {
+  config : config;
+  (* TODO: I am totally not sure about this type *)
+  acknowledges : (int, unit Lwt_condition.t) Base.Hashtbl.t;
+  mutable peers : Peer.t list;
+  mutable sequence_number : int;
+}
+
+(* Helper to increase the round *)
+let next_seq_no t =
+  let sequence_number = t.sequence_number in
+  t.sequence_number <- t.sequence_number + 1;
+  sequence_number
+
+let send_message client (recipient : Peer.t) message =
+  let message = Util.Encoding.pack bin_writer_message message in
+  Client.send_to client message recipient
+
+let send_ping_to client recipient = send_message client recipient Ping
+
+let send_acknowledge_to client recipient =
+  send_message client recipient Acknowledge
+
+let send_ping_request_to client recipient =
+  send_message client recipient (PingRequest recipient.socket_address)
+
+(* TODO: Totally not sure about this...
+   I don't know if I can get rid of it and only use `wait_ack_timeout` instead
+    But I feel like Lwt_condition allows ms to manage async thread and could be useful *)
+let[@warning "-32"] [@warning "-39"] rec wait_ack t sequence_number =
```

Comment:
Why this function is recursive? Are you sure you don't want to do anything after `Lwt_condition.wait` returns?

Reply:
Hum, I don't remember why I made it recursive at first... I just removed it.